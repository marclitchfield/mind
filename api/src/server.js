import { resolvers } from "./resolvers";
import { ApolloServer } from "apollo-server";
import { v1 as neo4j } from "neo4j-driver";
import { makeExecutableSchema } from "graphql-tools";
import dotenv from "dotenv";
import * as fs from 'fs';

export function createServer() {
  const typeDefs = fs.readdirSync(__dirname + '/schema')
    .map(file => fs.readFileSync(__dirname + '/schema/' + file, 'utf8'))
    .join('\n');

  const schema = makeExecutableSchema({
    typeDefs,
    resolvers,
    resolverValidationOptions: {
      requireResolversForResolveType: false
    }
  });

  dotenv.config();

  const driver = neo4j.driver(
    process.env.NEO4J_URI || "bolt://localhost:7687",
    neo4j.auth.basic(
      process.env.NEO4J_USER || "neo4j",
      process.env.NEO4J_PASSWORD || "password"
    )
  );

  return new ApolloServer({
    schema,
    context: {
      driver
    }
  });
}