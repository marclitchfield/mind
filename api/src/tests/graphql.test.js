import { createServer } from '../server';
import ApolloClient from "apollo-client";
import fetch from "node-fetch";
import { HttpLink } from "apollo-link-http";
import { InMemoryCache } from "apollo-cache-inmemory";
import { TestCases, TestCaseSetup } from './test-cases';
import gql from "graphql-tag";

const server = createServer();
let client;
const state = {};

beforeAll(async() => {
  client = new ApolloClient({
    link: new HttpLink({ uri: 'http://localhost:9999', fetch }),
    cache: new InMemoryCache()
  });
  
  await server.listen(9999);
  const response = await client.mutate({
    mutation: TestCaseSetup.mutation
  });

  updateState(response);
});

afterAll(async() => {
  await server.context.driver.close();
  await server.stop();
});

TestCases.forEach(testCase => {
  test(testCase.test, async() => {
    for(const mutation of testCase.mutations) {
      const mutateResponse = await client.mutate({ 
        mutation: applyStateReplacements(mutation),
      }).catch(handleClientError);
      updateState(mutateResponse);
    }
    
    const response = await client.query({ 
      query: applyStateReplacements(testCase.query),
    }).catch(handleClientError);
    
    const snapshot = JSON.parse(JSON.stringify(response));
    expect(snapshot).toMatchSnapshot();
  });
});

function updateState(response) {
  Object.keys(response.data).forEach(outerKey => {
    Object.keys(response.data[outerKey]).forEach(innerKey => {
      state[innerKey] = response.data[outerKey][innerKey];
    });
  });
}

function handleClientError(err) {
  if (err && err.networkError && err.networkError.result && err.networkError.result.errors) {
    const message = err.networkError.result.errors.map(formatError).join('\n');
    throw message;
  }
  throw err;
}

function formatError(graphqlError) {
  const locations = graphqlError.locations.map(loc => JSON.stringify(loc)).join('.');
  const message = `${graphqlError.extensions.code}: ${graphqlError.message} (${locations}`;
  return message.replace(/"/g, "'");
}

function applyStateReplacements(graphql) {
  let result = graphql.loc.source.body;
  const re = /\$\((.*?)\.(.*?)\)/g;
  let m;
  while (m = re.exec(result)) {
    if (m.length == 3) {
      result = result.replace(m[0], state[m[1]][m[2]]);
    }
  }
  return gql(result);
}

