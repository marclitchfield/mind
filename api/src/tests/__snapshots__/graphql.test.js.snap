// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection.addEvent 1`] = `
Object {
  "Collection": Object {
    "__typename": "Collection",
    "timeline": Array [
      Object {
        "__typename": "Event",
        "title": "Test",
      },
    ],
    "title": "Test",
  },
}
`;

exports[`Collection.clearLocation 1`] = `
Object {
  "Collection": Object {
    "__typename": "Collection",
    "location": null,
    "title": "Test",
  },
}
`;

exports[`Collection.createAtLocation 1`] = `
Object {
  "Space": Object {
    "__typename": "Space",
    "collections": Array [
      Object {
        "__typename": "Collection",
        "class": Object {
          "__typename": "Concept",
          "title": "Test",
        },
        "owners": Array [],
        "spaces": Array [
          Object {
            "__typename": "Space",
            "title": "Collection.createAtLocation",
          },
        ],
        "title": "Test",
      },
    ],
  },
}
`;

exports[`Collection.createForPerson 1`] = `
Object {
  "Space": Object {
    "__typename": "Space",
    "collections": Array [
      Object {
        "__typename": "Collection",
        "class": Object {
          "__typename": "Concept",
          "title": "Test",
        },
        "owners": Array [
          Object {
            "__typename": "Person",
            "title": "Test",
          },
        ],
        "spaces": Array [
          Object {
            "__typename": "Space",
            "title": "Collection.createForPerson",
          },
        ],
        "title": "Test",
      },
    ],
  },
}
`;

exports[`Collection.createInSpace 1`] = `
Object {
  "Space": Object {
    "__typename": "Space",
    "collections": Array [
      Object {
        "__typename": "Collection",
        "class": Object {
          "__typename": "Concept",
          "title": "Test",
        },
        "spaces": Array [
          Object {
            "__typename": "Space",
            "title": "Collection.createInSpace",
          },
        ],
        "title": "Test",
      },
    ],
  },
}
`;

exports[`Collection.createInstance 1`] = `
Object {
  "Space": Object {
    "__typename": "Space",
    "collections": Array [
      Object {
        "__typename": "Collection",
        "class": Object {
          "__typename": "Concept",
          "title": "Test",
        },
        "spaces": Array [
          Object {
            "__typename": "Space",
            "title": "Collection.createInstance",
          },
        ],
        "title": "Test",
      },
    ],
  },
}
`;

exports[`Collection.removeEvent 1`] = `
Object {
  "Collection": Object {
    "__typename": "Collection",
    "class": Object {
      "__typename": "Concept",
      "title": "Test",
    },
    "timeline": Array [],
    "title": "Test",
  },
}
`;

exports[`Collection.setClass 1`] = `
Object {
  "Collection": Object {
    "__typename": "Collection",
    "class": Object {
      "__typename": "Concept",
      "title": "Test2",
    },
    "title": "Test",
  },
}
`;

exports[`Collection.setLocation 1`] = `
Object {
  "Collection": Object {
    "__typename": "Collection",
    "class": Object {
      "__typename": "Concept",
      "title": "Test",
    },
    "location": Object {
      "__typename": "Location",
      "title": "Test2",
    },
    "title": "Test",
  },
}
`;

exports[`Concept.createInSpace 1`] = `
Object {
  "Space": Object {
    "__typename": "Space",
    "concepts": Array [
      Object {
        "__typename": "Concept",
        "spaces": Array [
          Object {
            "__typename": "Space",
            "title": "Concept.createInSpace",
          },
        ],
        "title": "Test",
      },
    ],
    "rootConcepts": Array [
      Object {
        "__typename": "Concept",
        "spaces": Array [
          Object {
            "__typename": "Space",
            "title": "Concept.createInSpace",
          },
        ],
        "title": "Test",
      },
    ],
  },
}
`;
